lib/builder.ts

```diff
@@ -98,74 +104,67 @@ class Builder extends EventEmitter {
     }
   }
 
+  build() {
+    // return this.buildPromise();
+    if (this._cancelationRequest) {
+      return Promise.reject(new BuilderError('Cannot start a build if one is already running'));
+    }
+    let promise = this._buildAsync();
+    this._cancelationRequest = new CancelationRequest(promise);
+    return promise;
+  }
+
   // Trigger a (re)build.
   //
   // Returns a promise that resolves when the build has finished. If there is a
   // build error, the promise is rejected with a Builder.BuildError instance.
   // This method will never throw, and it will never be rejected with anything
   // other than a BuildError.
-  build() {
-    if (this._cancelationRequest) {
-      return Promise.reject(new BuilderError('Cannot start a build if one is already running'));
-    }
-
-    let promise = Promise.resolve();
-
+  private async _buildAsync() {
     this.buildId++;
 
+    // Wipe all buildState objects at the beginning of the build
     for (const nw of this._nodeWrappers.values()) {
-      // Wipe all buildState objects at the beginning of the build
       nw.buildState = {};
+    }
+
+    await forceAsync(); // run the remaining code in the next tick.
 
-      promise = promise.then(() => {
-        let needsEndNode = false;
-        // We use a nested .then/.catch so that the .catch can only catch errors
-        // from this node, but not from previous nodes.
-        return promiseFinally(
-          Promise.resolve()
-            .then(() => this._cancelationRequest.throwIfRequested())
-            .then(() => {
-              this.emit('beginNode', nw);
-              needsEndNode = true;
-            })
-            .then(() => nw.build()),
-          () => {
-            if (needsEndNode) {
-              this.emit('endNode', nw);
-            }
+    try {
+      // Build each node
+      for (const nw of this._nodeWrappers.values()) {
+        this._cancelationRequest && this._cancelationRequest.throwIfRequested()
+        try {
+          try {
+            this.emit('beginNode', nw);
+            await nw.build();
+          } finally {
+            this.emit('endNode', nw);
           }
-        )
-          .then(() => this._cancelationRequest.throwIfRequested())
-          .catch(err => {
-            if (Cancelation.isCancelationError(err)) {
-              throw err;
-            } else {
-              throw new BuildError(err, nw);
-            }
-          });
-      });
-    }
+        } catch (err) {
+           // Not clear whether .build() is supposed to be able to throw a
+           // cancellation error or not but the promise version allowed it.
+          if (Cancelation.isCancelationError(err)) {
+            throw err;
+          } else {
+            throw new BuildError(err, nw);
+          }
+        }
+        this._cancelationRequest && this._cancelationRequest.throwIfRequested()
```

Comment:
Lines 153 and 136 don't seem like they are in the right position. The previous algorithm does the following.

1. `this._cancelationRequest.throwIfRequested()`
2. `nw.build()` 
3. `this._cancelationRequest.throwIfRequested()`
4. re-throw cancellation error (from 1, 3), or wrap error in a build error for improved build related diagnostics.

Reply:
I agree this is not a strict port of the promise code but I'm pretty certain these are functionally equivalent and I find my version to be more straightforward because it doesn't catch the cancellation error just to rethrow it.